### DISCUSSION TITLE
How can we check fingerprint is enabled or present in devices using service?

### DISCUSSION THREAD

**Discussion by dipakkumar1225** [appium/appium#15190] [Q&A]
Depending on fingerprint added in device or not, Fingerprint Access menu get displayed in the application. 
To automate this I am using command adb shell service call fingerprint 14. 
But I am getting result as:
Result: Parcel(
  0x00000000: ffffffff 00000083 0075004d 00740073 '........M.u.s.t.'
  0x00000010: 00680020 00760061 00200065 006e0061 ' .h.a.v.e. .a.n.'
  0x00000020: 00720064 0069006f 002e0064 00650070 'd.r.o.i.d...p.e.'
  0x00000030: 006d0072 00730069 00690073 006e006f 'r.m.i.s.s.i.o.n.'
  0x00000040: 0055002e 00450053 0042005f 004f0049 '..U.S.E._.B.I.O.'
  0x00000050: 0045004d 00520054 00430049 00700020 'M.E.T.R.I.C. .p.'
  0x00000060: 00720065 0069006d 00730073 006f0069 'e.r.m.i.s.s.i.o.'
  0x00000070: 002e006e 0020003a 0065004e 00740069 'n...:. .N.e.i.t.'
  0x00000080: 00650068 00200072 00730075 00720065 'h.e.r. .u.s.e.r.'
  0x00000090: 00320020 00300030 00200030 006f006e ' .2.0.0.0. .n.o.'
  0x000000a0: 00200072 00750063 00720072 006e0065 'r. .c.u.r.r.e.n.'
  0x000000b0: 00200074 00720070 0063006f 00730065 't. .p.r.o.c.e.s.'
  0x000000c0: 00200073 00610068 00200073 006e0061 's. .h.a.s. .a.n.'
  0x000000d0: 00720064 0069006f 002e0064 00650070 'd.r.o.i.d...p.e.'
  0x000000e0: 006d0072 00730069 00690073 006e006f 'r.m.i.s.s.i.o.n.'
  0x000000f0: 0055002e 00450053 0042005f 004f0049 '..U.S.E._.B.I.O.'
  0x00000100: 0045004d 00520054 00430049 0000002e 'M.E.T.R.I.C.....'
  0x00000110: 000003b4 000001d5 00610009 00200074 '..........a.t. .'
  0x00000120: 006e0061 00720064 0069006f 002e0064 'a.n.d.r.o.i.d...'
  0x00000130: 00700061 002e0070 006f0043 0074006e 'a.p.p...C.o.n.t.'
  0x00000140: 00780065 00490074 0070006d 002e006c 'e.x.t.I.m.p.l...'
  0x00000150: 006e0065 006f0066 00630072 00280065 'e.n.f.o.r.c.e.(.'
  0x00000160: 006f0043 0074006e 00780065 00490074 'C.o.n.t.e.x.t.I.'
  0x00000170: 0070006d 002e006c 0061006a 00610076 'm.p.l...j.a.v.a.'
  0x00000180: 0032003a 00310030 00290038 0009000a ':.2.0.1.8.).....'
  0x00000190: 00740061 00610020 0064006e 006f0072 'a.t. .a.n.d.r.o.'
  0x000001a0: 00640069 0061002e 00700070 0043002e 'i.d...a.p.p...C.'
  0x000001b0: 006e006f 00650074 00740078 006d0049 'o.n.t.e.x.t.I.m.'
  0x000001c0: 006c0070 0065002e 0066006e 0072006f 'p.l...e.n.f.o.r.'
  0x000001d0: 00650063 00610043 006c006c 006e0069 'c.e.C.a.l.l.i.n.'
  0x000001e0: 004f0067 00530072 006c0065 00500066 'g.O.r.S.e.l.f.P.'
  0x000001f0: 00720065 0069006d 00730073 006f0069 'e.r.m.i.s.s.i.o.'
  0x00000200: 0028006e 006f0043 0074006e 00780065 'n.(.C.o.n.t.e.x.'
  0x00000210: 00490074 0070006d 002e006c 0061006a 't.I.m.p.l...j.a.'
  0x00000220: 00610076 0032003a 00340030 00290036 'v.a.:.2.0.4.6.).'
  0x00000230: 0009000a 00740061 00630020 006d006f '....a.t. .c.o.m.'
  0x00000240: 0061002e 0064006e 006f0072 00640069 '..a.n.d.r.o.i.d.'
  0x00000250: 0073002e 00720065 00650076 002e0072 '..s.e.r.v.e.r...'
  0x00000260: 00690062 006d006f 00740065 00690072 'b.i.o.m.e.t.r.i.'
  0x00000270: 00730063 0042002e 006f0069 0065006d 'c.s...B.i.o.m.e.'
  0x00000280: 00720074 00630069 00650053 00760072 't.r.i.c.S.e.r.v.'
  0x00000290: 00630069 00420065 00730061 002e0065 'i.c.e.B.a.s.e...'
  0x000002a0: 00680063 00630065 0050006b 00720065 'c.h.e.c.k.P.e.r.'
  0x000002b0: 0069006d 00730073 006f0069 0028006e 'm.i.s.s.i.o.n.(.'
  0x000002c0: 00690042 006d006f 00740065 00690072 'B.i.o.m.e.t.r.i.'
  0x000002d0: 00530063 00720065 00690076 00650063 'c.S.e.r.v.i.c.e.'
  0x000002e0: 00610042 00650073 006a002e 00760061 'B.a.s.e...j.a.v.'
  0x000002f0: 003a0061 00320031 00340032 000a0029 'a.:.1.2.2.4.)...'
  0x00000300: 00610009 00200074 006f0063 002e006d '..a.t. .c.o.m...'
  0x00000310: 006e0061 00720064 0069006f 002e0064 'a.n.d.r.o.i.d...'
  0x00000320: 00650073 00760072 00720065 0062002e 's.e.r.v.e.r...b.'
  0x00000330: 006f0069 0065006d 00720074 00630069 'i.o.m.e.t.r.i.c.'
  0x00000340: 002e0073 00690066 0067006e 00720065 's...f.i.n.g.e.r.'
  0x00000350: 00720070 006e0069 002e0074 00690046 'p.r.i.n.t...F.i.'
  0x00000360: 0067006e 00720065 00720070 006e0069 'n.g.e.r.p.r.i.n.'
  0x00000370: 00530074 00720065 00690076 00650063 't.S.e.r.v.i.c.e.'
  0x00000380: 0063002e 00650068 006b0063 00730055 '..c.h.e.c.k.U.s.'
  0x00000390: 00420065 006f0069 0065006d 00720074 'e.B.i.o.m.e.t.r.'
  0x000003a0: 00630069 00650050 006d0072 00730069 'i.c.P.e.r.m.i.s.'
  0x000003b0: 00690073 006e006f 00460028 006e0069 's.i.o.n.(.F.i.n.'
  0x000003c0: 00650067 00700072 00690072 0074006e 'g.e.r.p.r.i.n.t.'
  0x000003d0: 00650053 00760072 00630069 002e0065 'S.e.r.v.i.c.e...'
  0x000003e0: 0061006a 00610076 0038003a 00300035 'j.a.v.a.:.8.5.0.'
  0x000003f0: 000a0029 00610009 00200074 006f0063 ').....a.t. .c.o.'
  0x00000400: 002e006d 006e0061 00720064 0069006f 'm...a.n.d.r.o.i.'
  0x00000410: 002e0064 00650073 00760072 00720065 'd...s.e.r.v.e.r.'
  0x00000420: 0062002e 006f0069 0065006d 00720074 '..b.i.o.m.e.t.r.'
  0x00000430: 00630069 002e0073 00690042 006d006f 'i.c.s...B.i.o.m.'
  0x00000440: 00740065 00690072 00530063 00720065 'e.t.r.i.c.S.e.r.'
  0x00000450: 00690076 00650063 00610042 00650073 'v.i.c.e.B.a.s.e.'
  0x00000460: 0063002e 006e0061 00730055 00420065 '..c.a.n.U.s.e.B.'
  0x00000470: 006f0069 0065006d 00720074 00630069 'i.o.m.e.t.r.i.c.'
  0x00000480: 00420028 006f0069 0065006d 00720074 '(.B.i.o.m.e.t.r.'
  0x00000490: 00630069 00650053 00760072 00630069 'i.c.S.e.r.v.i.c.'
  0x000004a0: 00420065 00730061 002e0065 0061006a 'e.B.a.s.e...j.a.'
  0x000004b0: 00610076 0031003a 00300030 00290034 'v.a.:.1.0.0.4.).'
  0x000004c0: 0000000a                            '....            ')
  
  Please let me know what I am missing or any other workaround?

**Comment by KazuCocoa** [comment]
Instead of https://appium.io/docs/en/commands/device/authentication/finger-print/ ?

Probably the command returned https://developer.android.com/reference/android/os/Parcel bytecode

**Reply by dipakkumar1225** [reply]
Thanks for the reply

Hope below image will help to explain my requirement.
I want to validate, is Fingerprint menu displayed when fingerprint present/added in device? 
Menu when fingerprint is not present.
![image](https://user-images.githubusercontent.com/46619348/111060664-9436ae00-84c4-11eb-85ee-54752f01244d.png)

Menu after fingerprint added. 
![image](https://user-images.githubusercontent.com/46619348/111060677-aadd0500-84c4-11eb-8f6f-94d12025baf1.png)

I am trying to achieve the same using adb shell service command as explained [here](https://stackoverflow.com/questions/20227326/where-to-find-info-on-androids-service-call-shell-command), 

**service call service_name number_at_which_function_appears_in_service_name.aidl argument_type_like i32 or i64**

https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android11-mainline-release/core/java/android/hardware/fingerprint/IFingerprintService.aidl

adb shell service command
adb shell service call fingerprint 14 i64 1 s16 "USE_BIOMETRIC"

Please let me know what I missed. or is there any other way to check the same.

**Reply by KazuCocoa** [reply]
So, you were able to achieve it when you call the adb shell service directly to the device but could not work via Appium's adb shell command by http://appium.io/docs/en/writing-running-appium/android/android-shell/ ?

If it did not work on the direct method as well, Appium cannot.
Then, I have no good knowledge about it.
